### PR TITLE
[CALCITE-5416] RelToSql converter generates invalid code when merging…

### PR DESCRIPTION
**Pull request by JiajunBernoulli**
… rollup and sort clauses

### CODE REVIEW

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java [hunk 1]
```diff
@@ -672,9 +672,11 @@ private static String toSql(RelNode root, SqlDialect dialect,
         + "FROM \"foodmart\".\"product\"\n"
         + "GROUP BY ROLLUP(\"product_class_id\", \"brand_name\")\n"
         + "ORDER BY \"product_class_id\", \"brand_name\"";
-    final String expectedMysql = "SELECT `product_class_id`, `brand_name`\n"
+    final String expectedMysql = "SELECT *\n"
```
Comment (LeonidChistov) [inline core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:675]: That is probably excessive - if order of grouping fields matches the order of sort fields, we can omit ORDER BY clause (as it was done in the old solution).

Comment (JiajunBernoulli) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java]: I restored to old solution. Thanks for your review.

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java [hunk 2]
```diff
@@ -868,19 +867,35 @@ public Result visit(Sort e) {
...
+            || isImplicitlySort) {
+          offsetFetch(e, builder);
+          return result;
+        }
+        // It does not allow "WITH ROLLUP" in combination with "ORDER BY",
+        // so generate the grouped result apply ORDER BY to it.
+        SqlSelect sqlSelect = result.subSelect();
+        SqlNodeList sortExps = exprList(builder.context, e.getSortExps());
+        sqlSelect.setOrderBy(sortExps);
+        SqlNode offset = e.offset == null ? null : builder.context.toSql(null, e.offset);
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java:894]: for offset and fetch I think `if` would be clearer than `?`

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java [hunk 3]
```diff
@@ -868,19 +867,35 @@ public Result visit(Sort e) {
...
-        final Set<Integer> groupList = new LinkedHashSet<>();
-        for (RelFieldCollation fc : e.collation.getFieldCollations()) {
-          groupList.add(aggregate.getGroupSet().nth(fc.getFieldIndex()));
-        }
-        groupList.addAll(Aggregate.Group.getRollup(aggregate.getGroupSets()));
+        boolean isImplicitlySort = rollupList.subList(0, sortList.size()).equals(sortList);
         final Builder builder =
-            visitAggregate(aggregate, ImmutableList.copyOf(groupList),
+            visitAggregate(aggregate,
+                rollupList,
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java:881]: put `rollupList` on previous line

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java [hunk 4]
```diff
@@ -868,19 +867,35 @@ public Result visit(Sort e) {
...
+            .map(f -> aggregate.getGroupSet().nth(f.getFieldIndex()))
+            .collect(Collectors.toList());
+        // "GROUP BY x, y WITH ROLLUP" implicitly sorts by x, y,
         // so skip the ORDER BY.
-        final Set<Integer> groupList = new LinkedHashSet<>();
-        for (RelFieldCollation fc : e.collation.getFieldCollations()) {
-          groupList.add(aggregate.getGroupSet().nth(fc.getFieldIndex()));
-        }
-        groupList.addAll(Aggregate.Group.getRollup(aggregate.getGroupSets()));
+        boolean isImplicitlySort = rollupList.subList(0, sortList.size()).equals(sortList);
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java:878]: should be `final`

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java [hunk 5]
```diff
@@ -868,19 +867,35 @@ public Result visit(Sort e) {
...
-        offsetFetch(e, builder);
-        return builder.result();
+        Result result = builder.result();
+        if (sortList.isEmpty()
+            || isImplicitlySort) {
+          offsetFetch(e, builder);
+          return result;
+        }
+        // It does not allow "WITH ROLLUP" in combination with "ORDER BY",
+        // so generate the grouped result apply ORDER BY to it.
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java:890]: Change "It" to "MySQL"

**Review by julianhyde — CHANGES_REQUESTED** [review]
LGTM, with the suggestions I made.

Also, improve the commit message and bug summary. Mention MySQL, JDBC adapter, GROUP BY WITH ROLLUP, ORDER BY.

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java (hunk 4 above)
Comment (JiajunBernoulli) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: I didn't have this consciousness before, thank you for letting me learn.

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java (hunk 2 above)
Comment (JiajunBernoulli) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: Yes, `if` is clearer.

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java (hunk 3 above)
Comment (JiajunBernoulli) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: Ok

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java (hunk 5 above)
Comment (JiajunBernoulli) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java:889]: done

**Comment by JiajunBernoulli** [comment]
> LGTM, with the suggestions I made.
> 
> Also, improve the commit message and bug summary. Mention MySQL, JDBC adapter, GROUP BY WITH ROLLUP, ORDER BY.

Thanks for your review. 

> LGTM, with the suggestions I made.
> 
> Also, improve the commit message and bug summary. Mention MySQL, JDBC adapter, GROUP BY WITH ROLLUP, ORDER BY.

Thanks for your review!

**Pull request closed without merging**